# Задание по программированию: Выбор семейства распределений в наивном байесе

В этом задании вы сможете немного расслабиться после реализации случайного леса и градиентного бустинга по мотивам материалов прошлой недели. Все, что нужно будет делать — запускать методы из sklearn. Вам предлагается выяснить, какое распределение лучше использовать в наивном байесовском классификаторе в зависимости от вида признаков.

Загрузите датасеты **digits** и **breast_cancer** из sklearn.datasets. Выведите несколько строчек из обучающих выборок и посмотрите на признаки. С помощью **sklearn.cross_validation.cross_val_score** c настройками по умолчанию и вызова метода **mean()** у возвращаемого этой функцией numpy.ndarray, сравните качество работы наивных байесовских классификаторов на этих двух датасетах. Для сравнения предлагается использовать **BernoulliNB, MultinomialNB и GaussianNB**. Насколько полученные результаты согласуются с рекомендациями из лекций?

Два датасета, конечно, еще не повод делать далеко идущие выводы, но при желании вы можете продолжить исследование на других выборках (например, из UCI репозитория).

In [9]:
from sklearn import datasets
from sklearn.model_selection import cross_validate, cross_val_score
digits = datasets.load_digits();
breast_cancer = datasets.load_breast_cancer();

In [10]:
from sklearn.naive_bayes import  BernoulliNB, MultinomialNB, GaussianNB

In [26]:
def estimate_quality(classifier, data):
    resultCV = cross_val_score(classifier, data.data, data.target, cv=3, n_jobs = 3)
    return resultCV.mean()
    
    
print("Estimations for digits dataset")
print("BernoulliNB", estimate_quality (BernoulliNB(), digits))
print("MultinomialNB", estimate_quality (MultinomialNB(), digits))
print("GaussianNB", estimate_quality (GaussianNB(), digits))
print()
print("Estimations for breast_cancer dataset")
print("BernoulliNB", estimate_quality (BernoulliNB(), breast_cancer))
print("MultinomialNB", estimate_quality (MultinomialNB(), breast_cancer))
print("GaussianNB", estimate_quality (GaussianNB(), breast_cancer))




Estimations for digits dataset
BernoulliNB 0.8235948803561491
MultinomialNB 0.870895937673901
GaussianNB 0.8135781858653311

Estimations for breast_cancer dataset
BernoulliNB 0.6274204028589994
MultinomialNB 0.8945790401930752
GaussianNB 0.9367492806089297


Для сдачи задания, ответьте на приведенные ниже вопросы.

**Вопрос 1**

Каким получилось максимальное качество классификации на датасете breast_cancer?
**Ответ: 0.93**

**Вопрос 2**

Каким получилось максимальное качество классификации на датасете digits?
**Ответ: 0.87**

**Вопрос 3**  
digits - целочисленные признаки  
breast_cancer - вещественные признаки  
Выберите верные утверждения и запишите их номера через пробел (в порядке возрастания номера):

    1) На вещественных признаках лучше всего сработал наивный байесовский классификатор с распределением Бернулли - [Нет]

    2) На вещественных признаках лучше всего сработал наивный байесовский классификатор с мультиномиальным распределением - [Нет]

    3) Мультиномиальное распределение лучше показало себя на выборке с целыми неотрицательными значениями признаков - [Да]

    4) На вещественных признаках лучше всего сработало нормальное распределение - [Да]